shared/src/main/scala/mlscript/JSBackend.scala

```diff
@@ -407,123 +244,157 @@ class JSBackend {
+              translateTerm(term)(topLevelScope) :: Nil
+            ).stmt :: Nil
+        })
     val epilogue = resultsIdent.member("map")(JSIdent(prettyPrinterName)).`return` :: Nil
     JSImmEvalFn(Nil, R(polyfill.emit() ::: stmts ::: epilogue), Nil).toSourceCode.toLines
   }
 }
 
 class JSTestBackend extends JSBackend {
-  private val resultName = topLevelScope allocateJavaScriptName "res"
+  private val lastResultSymbol = topLevelScope declareValue "res"
+  private val resultIdent = JSIdent(lastResultSymbol.runtimeName)
 
   private var numRun = 0
 
-  // TODO: make the return type more readable
-  def apply(pgrm: Pgrm): (Str, Bool) \/ TestCode = try {
-    R(generate(pgrm))
-  } catch {
-    case e: CodeGenError => L(e.getMessage() -> false) // Intended errors
-    case e: Throwable => L(e.getMessage() -> true) // Unexpected crashes
-  }
+  /**
+    * Generate a piece of code for test purpose. It can be invoked repeatedly.
+    */
+  def apply(pgrm: Pgrm, allowEscape: Bool): JSTestBackend.Result =
+    try generate(pgrm)(topLevelScope, allowEscape) catch {
+      case e: CodeGenError => JSTestBackend.IllFormedCode(e.getMessage())
+      case e: UnimplementedError => JSTestBackend.Unimplemented(e.getMessage())
+      case e: Throwable => JSTestBackend.UnexpectedCrash(e.getMessage())
+    }
 
-  // Generate code for test.
-  private def generate(pgrm: Pgrm): TestCode = {
+  /**
+    * Generate JavaScript code which targets MLscript test from the given block.
+    *
+    * @param pgrm the program to translate
+    * @param scope the top-level scope
+    * @param allowEscape whether to try executing code even if it refers to unimplemented definitions
+    * @return
+    */
+  private def generate(pgrm: Pgrm)(implicit scope: Scope, allowEscape: Bool): JSTestBackend.TestCode = {
     val (diags, (typeDefs, otherStmts)) = pgrm.desugared
 
-    // Collect type aliases into a map so we can normalize them.
-    typeDefs foreach {
-      case TypeDef(Als, TypeName(name), tparams, body, _, _) =>
-        val tnames = tparams map { case TypeName(nme) => nme }
-        typeAliasMap(name) = tnames -> body
-      case TypeDef(Trt, TypeName(name), _, _, _, _) =>
-        traitNames += name
-      case _ => ()
-    }
-
-    val resultIdent = JSIdent(resultName)
-
-    // Generate hoisted declarations.
-    val defStmts = typeDefs
-      .flatMap { case TypeDef(kind, TypeName(name), typeParams, actualType, _, mthDefs) =>
-        kind match {
-          case Cls =>
-            topLevelScope declare name
-            classNames += name
-            val cls = translateClassDeclaration(name, actualType, mthDefs)(topLevelScope)
-            nameClsMap += name -> cls
-            S(cls)
-          case Trt => N
-          case Als => N
-        }
-      }
+    val classSymbols = declareTypeDefs(typeDefs)
+    val defStmts = sortClassSymbols(classSymbols).map { case (derived, base) =>
+      translateClassDeclaration(derived, base)(topLevelScope)
+    }.toList
```

Comment:
After we got all sorted newly-defined classes (together with their base classes!), we pass them to `translateClassDeclaration`, which is the second consumer of base classes.